contracts/atomic-swap/src/contract.rs

```diff
@@ -0,0 +1,619 @@
+    }
+
+    if env.message.sent_funds.is_empty() {
+        return Err(StdError::generic_err(
+            "Send some coins to create an atomic swap",
+        ));
+    }
+
+    // Ensure this is 32 bytes hex-encoded
+    let _ = parse_hex_32(&msg.hash)?;
+
+    let recipient_raw = deps.api.canonical_address(&msg.recipient)?;
+
+    let swap = AtomicSwap {
+        hash: msg.hash.clone(),
+        recipient: recipient_raw,
+        source: deps.api.canonical_address(&env.message.sender)?,
+        end_height: msg.end_height,
+        end_time: msg.end_time,
+        balance: env.message.sent_funds.clone(),
+    };
+
+    if swap.is_expired(&env) {
+        return Err(StdError::generic_err("Expired atomic swap"));
+    }
+
+    // Try to store it, fail if the id already exists (unmodifiable swaps)
+    atomic_swaps(&mut deps.storage).update(msg.id.as_bytes(), |existing| match existing {
+        None => Ok(swap),
+        Some(_) => Err(StdError::generic_err("Atomic swap already exists")),
+    })?;
+
+    let mut res = HandleResponse::default();
+    res.log = vec![
+        log("action", "create"),
+        log("id", msg.id),
+        log("hash", msg.hash),
+        log("recipient", msg.recipient),
+    ];
+    Ok(res)
+}
+
+pub fn try_release<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    id: String,
+    preimage: String,
+) -> StdResult<HandleResponse> {
+    let swap = atomic_swaps_read(&deps.storage).load(id.as_bytes())?;
+    if swap.is_expired(&env) {
+        return Err(StdError::generic_err("Atomic swap expired"));
+    }
+
+    let hash = Sha256::digest(&parse_hex_32(&preimage)?);
+    let expected = parse_hex_32(&swap.hash)?;
+    if hash.as_slice() != expected.as_slice() {
+        return Err(StdError::generic_err("Invalid preimage"));
+    }
+
+    let rcpt = deps.api.human_address(&swap.recipient)?;
+
+    // We delete the swap
+    atomic_swaps(&mut deps.storage).remove(id.as_bytes());
+
+    // Send all tokens out
+    let msgs = send_native_tokens(&env.contract.address, &rcpt, swap.balance);
+    Ok(HandleResponse {
+        messages: msgs,
+        log: vec![
+            log("action", "release"),
+            log("id", id),
+            log("preimage", preimage),
+            log("to", rcpt),
+        ],
+        data: None,
+    })
+}
+
+pub fn try_refund<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    id: String,
+) -> StdResult<HandleResponse> {
+    let swap = atomic_swaps_read(&deps.storage).load(id.as_bytes())?;
+    // Anyone can try to refund, as long as the contract is expired
+    if !swap.is_expired(&env) {
+        return Err(StdError::generic_err("Atomic swap not yet expired"));
+    }
+
+    let rcpt = deps.api.human_address(&swap.source)?;
+
+    // We delete the swap
+    atomic_swaps(&mut deps.storage).remove(id.as_bytes());
+
+    let msgs = send_native_tokens(&env.contract.address, &rcpt, env.message.sent_funds);
```

Comment:
Nope. We do not send back the user the coins they sent in this transaction. We need to return the values stored in the swap.

`env.message.sent_funds` => `swap.balance` (just like release above)

Reply:
Oops. Good catch!

Reply:
I don't know what I was thinking. And, it's unfortunate that the broken sender balance in unit tests didn't flag this.